teach ivalue about List[Optional[Tensor]], fix fallbacks

bdhirsh opened the pull request:
IValue's treat the `[Optional[Tensor]]` type a bit weird:
- For normal `c10::optional<Tensor>` types, the optional-ness gets folded away into an undefined tensor. So boxed fallbacks don't need to handle the `c10::optional<Tensor>` case directly ([ivalue code](https://github.com/pytorch/pytorch/blob/c2d5f6a5a43ae3fd8ce78c88964c1357b94537b3/aten/src/ATen/core/ivalue_inl.h#L2020))
- For `c10::list<c10::optional<Tensor>>` though, boxed fallbacks need to handle that directly.

I added API's for checking / retrieving `OptionalTensorList` from an `IValue`, and also updated the functionalization + lazy_tensor_cpu boxed fallbacks to correctly handle them. 

I also added a test for `index_put_` (which used to fail, because once of its arguments was a list of optional tensors).



Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527 [prototype] integrate functionalization <> LTC torchscript backend
* #75913 functionalization: introduce a "zero()" aten op
* #75819 functionalization: avoid some unnecessary view_copy calls
* #75717 fix unfold for meta tensors
* **#75716 teach ivalue about List[Optional[Tensor]], fix fallbacks**
* #75818 fix out= op handling for functionalization
* #75302 split out functionalization codegen to use view_copy operators

Differential Revision: [D35705376](https://our.internmc.facebook.com/intern/diff/D35705376)

bdhirsh committed:
teach ivalue about List[Optional[Tensor]], fix fallbacks

[ghstack-poisoned]

bdhirsh committed:
Update on "teach ivalue about List[Optional[Tensor]], fix fallbacks"




[ghstack-poisoned]

bdhirsh committed:
Update on "teach ivalue about List[Optional[Tensor]], fix fallbacks"




[ghstack-poisoned]

bdhirsh committed:
Update on "teach ivalue about List[Optional[Tensor]], fix fallbacks"

IValue's treat the `[Optional[Tensor]]` type a bit weird:
- For normal `c10::optional<Tensor>` types, the optional-ness gets folded away into an undefined tensor. So boxed fallbacks don't need to handle the `c10::optional<Tensor>` case directly ([ivalue code](https://github.com/pytorch/pytorch/blob/c2d5f6a5a43ae3fd8ce78c88964c1357b94537b3/aten/src/ATen/core/ivalue_inl.h#L2020))
- For `c10::list<c10::optional<Tensor>>` though, boxed fallbacks need to handle that directly.

I added API's for checking / retrieving `OptionalTensorList` from an `IValue`, and also updated the functionalization + lazy_tensor_cpu boxed fallbacks to correctly handle them. 

I also added a test for `index_put_` (which used to fail, because once of its arguments was a list of optional tensors).






[ghstack-poisoned]

bdhirsh commented on torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp:
```diff
@@ -187,9 +226,14 @@ void ts_eager_fallback(
+      auto eager_ivalue = c10::IValue(c10::List<c10::optional<at::Tensor>>(
+          to_eager(ivalue.toOptionalTensorVector(), device_type)));
+      (*stack)[arguments_begin + idx] = std::move(eager_ivalue);
+      opt_tensorlist_args.push_back(ivalue.toOptionalTensorList());
```
cc @wconstab, I think this should fix `index_put_` for LTC

wconstab replied:
I forgot (if I knew) what the problem with index_put_ was for LTC.  Is this fixing something that is broken when index_put falls back?

zou3519 commented on aten/src/ATen/FunctionalTensorWrapper.h:
```diff
@@ -124,7 +123,9 @@ TORCH_API bool isFunctionalTensor(const c10::List<c10::optional<Tensor>>& t_list
 TORCH_API bool isFunctionalTensor(const c10::ArrayRef<Tensor> t_list);
 
 TORCH_API Tensor to_functional_tensor(const Tensor& tensor);
+TORCH_API c10::optional<Tensor> to_functional_tensor(const c10::optional<Tensor>& tensor);
```
How does C++ pick which overload to use if you pass in to_functional_tensor(tensor)? (this doesn't really matter because the code compiles but I'm curious)

zou3519 commented on aten/src/ATen/FunctionalTensorWrapper.h:
```diff
@@ -124,7 +123,9 @@ TORCH_API bool isFunctionalTensor(const c10::List<c10::optional<Tensor>>& t_list
 TORCH_API Tensor to_functional_tensor(const Tensor& tensor);
+TORCH_API c10::optional<Tensor> to_functional_tensor(const c10::optional<Tensor>& tensor);
 TORCH_API c10::List<Tensor> to_functional_tensor(const c10::List<Tensor>& t_list);
+TORCH_API c10::List<c10::optional<Tensor>> to_functional_tensor(const c10::List<c10::optional<Tensor>>& t_list);
```
(no action required) We really need something like a c++ tree_map for all dispatcher types

zou3519 commented on aten/src/ATen/FunctionalizeFallbackKernel.cpp:
```diff
@@ -34,6 +34,15 @@ namespace {
+      } else if (ivalue.isOptionalTensorList()) {
+        any_tensor_inputs = true;
+        auto opt_tensors = ivalue.toOptionalTensorList();
+        if (at::functionalization::impl::isFunctionalTensor(opt_tensors)) {
```
Is there an invariant that either all or none of these must be functional tensors, or can there be some hybrid mix?

zou3519 commented on torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp:
```diff
@@ -68,6 +69,35 @@ std::vector<at::Tensor> to_eager(
+std::vector<c10::optional<at::Tensor>> to_eager(
+    const std::vector<c10::optional<at::Tensor>>& tensors,
+    c10::DeviceType device_type) {
+  // We can't just call _to_eager() on the entire list of Tensors because it
+  // will break on undefined tensors. Separate out undefined tensors first.
```
I'm confused -- std::vector<c10::optional<at::Tensor>> probably came from an IValue that is `List[Optional[Tensor]]`, but it can contain undefined tensors?

ezyang replied:
it's just copy pasted from the impl above lol

ezyang replied:
I think it should be possible to assume that inside optional they are never undefined, and things should work

ezyang commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -435,6 +435,14 @@ bool IValue::isTensorList() const {
+    return false;
+  }
+  const auto& ty = static_cast<detail::ListImpl*>(payload.u.as_intrusive_ptr)->elementType;
+  return ty == c10::getTypePtr<c10::optional<at::Tensor>>();
```
How come `isListOf` can't be used here

ezyang reviewed (approved):
okey dokey. Check Richard's comments

bdhirsh committed:
Update on "teach ivalue about List[Optional[Tensor]], fix fallbacks"

IValue's treat the `[Optional[Tensor]]` type a bit weird:
- For normal `c10::optional<Tensor>` types, the optional-ness gets folded away into an undefined tensor. So boxed fallbacks don't need to handle the `c10::optional<Tensor>` case directly ([ivalue code](https://github.com/pytorch/pytorch/blob/c2d5f6a5a43ae3fd8ce78c88964c1357b94537b3/aten/src/ATen/core/ivalue_inl.h#L2020))
- For `c10::list<c10::optional<Tensor>>` though, boxed fallbacks need to handle that directly.

I added API's for checking / retrieving `OptionalTensorList` from an `IValue`, and also updated the functionalization + lazy_tensor_cpu boxed fallbacks to correctly handle them. 

I also added a test for `index_put_` (which used to fail, because once of its arguments was a list of optional tensors).






[ghstack-poisoned]

bdhirsh committed:
Update on "teach ivalue about List[Optional[Tensor]], fix fallbacks"

IValue's treat the `[Optional[Tensor]]` type a bit weird:
- For normal `c10::optional<Tensor>` types, the optional-ness gets folded away into an undefined tensor. So boxed fallbacks don't need to handle the `c10::optional<Tensor>` case directly ([ivalue code](https://github.com/pytorch/pytorch/blob/c2d5f6a5a43ae3fd8ce78c88964c1357b94537b3/aten/src/ATen/core/ivalue_inl.h#L2020))
- For `c10::list<c10::optional<Tensor>>` though, boxed fallbacks need to handle that directly.

I added API's for checking / retrieving `OptionalTensorList` from an `IValue`, and also updated the functionalization + lazy_tensor_cpu boxed fallbacks to correctly handle them. 

I also added a test for `index_put_` (which used to fail, because once of its arguments was a list of optional tensors).






[ghstack-poisoned]

bdhirsh committed:
Update on "teach ivalue about List[Optional[Tensor]], fix fallbacks"

IValue's treat the `[Optional[Tensor]]` type a bit weird:
- For normal `c10::optional<Tensor>` types, the optional-ness gets folded away into an undefined tensor. So boxed fallbacks don't need to handle the `c10::optional<Tensor>` case directly ([ivalue code](https://github.com/pytorch/pytorch/blob/c2d5f6a5a43ae3fd8ce78c88964c1357b94537b3/aten/src/ATen/core/ivalue_inl.h#L2020))
- For `c10::list<c10::optional<Tensor>>` though, boxed fallbacks need to handle that directly.

I added API's for checking / retrieving `OptionalTensorList` from an `IValue`, and also updated the functionalization + lazy_tensor_cpu boxed fallbacks to correctly handle them. 

I also added a test for `index_put_` (which used to fail, because once of its arguments was a list of optional tensors).






[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35705376).
